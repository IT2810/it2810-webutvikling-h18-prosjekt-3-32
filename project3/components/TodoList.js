import React from 'react';
import { Button, AsyncStorage, ScrollView, TextInput, TouchableOpacity, Text, View } from 'react-native';
import ListItem from "./ListItem.js";
import styles from "../stylesheets/TodoList.style.js";

export default class TodoList extends React.Component {
    constructor(props){
        super(props);

        //State with current todonr, which increases, and the current todotext that the user is entering
        this.state = {
            currTodoNr: 1,
            todoText: "",
            showFinishedTodos: true,
            todoList: [],
            finishedTodoList:[],
            showList:[],
        };

        //Binding functions
        this.updateTodoList = this.updateTodoList.bind(this);
        this.updateShowList = this.updateShowList.bind(this);
        this.addTodo = this.addTodo.bind(this);
        this.updateSortedList = this.updateSortedList.bind(this);
        this.handleFinishedTodo = this.handleFinishedTodo.bind(this);
        this.deleteTodo = this.deleteTodo.bind(this);
        this.myTextInput = React.createRef();
    };

    componentDidMount(){
        let todoList = [];
        let finishedTodoList = [];
        //Gets the highest todoNr and updates the state to no todos have the same todoNr and key.
        AsyncStorage.getAllKeys((err, keys) => {
            AsyncStorage.multiGet(keys, (err, stores) => {
                stores.map((result, i, store) => {
                    // get at each store's key/value so you can work with it
                    let key = store[i][0];
                    let value = store[i][1];
                    if(key === "CurrentTodoNr"){
                        this.setState({currTodoNr:parseInt(value)+1});
                    }

                    //Checks if the item is a to-do
                    if(key.substring(0,4) === "todo"){
                        const valueList = JSON.parse(value);
                        todoList.push({
                            key: key,
                            todoNr: key,
                            todoText: valueList[0],
                            todoDate: valueList[1],
                            done: valueList[2],
                        });
                    }

                    //Checks if the item is a finished to-do
                    else if (key.substring(0,4) === "done"){
                        const valueList = JSON.parse(value);
                        finishedTodoList.push({
                            key: key,
                            todoNr: key,
                            todoText: valueList[0],
                            done: valueList[2],
                        });
                    }
                });

                //Sorts the imported list from AsyncStorage and updates the state to all three lists
                const sortedByDateTodos = this.sortByDate(todoList);
                const sortedByDateFinishedTodos = this.sortByDate(finishedTodoList);
                this.setState({
                    showList:sortedByDateTodos,
                    todoList:sortedByDateTodos,
                    finishedTodoList:sortedByDateFinishedTodos,
                });
            });
        });
    }

    //addTodo adds a new to-do, add it to a list, and passes the list to app.js, since the to-do-list is needed at the home screen.
    //Each element in addTodo gets a key, todoNr, a name and a date(which is empty until the users sets a due date)
    addTodo = (id) => {
        this.myTextInput.current.clear();
        if(this.state.todoText !== ""){
            //Creates a list, sliced from the current todolist
            //Adds the new to-do to the list.
            const addList = this.state.todoList.slice();
            addList.push({
                key: id,
                todoNr: "todo"+id,
                todoText: this.state.todoText,
                todoDate: "",
                done: false,
            });
            this.setState({
                currTodoNr: this.state.currTodoNr + 1,
                todoText : "",
            });

            //The list is being sat to both to-do- and showlist for immediate change on the screen.
            this.updateTodoList(addList);
            this.updateShowList(addList);

            //The new to-do is also added in the AsyncStorage, aswell as the latest todoNr.
            const AsyncList = [this.state.todoText, "", false];
            this.storeTodo("todo"+id.toString(), JSON.stringify(AsyncList));
            this.storeTodo("CurrentTodoNr", id.toString());
        }
        else{alert("The todo needs a description!!");}
    };

    //This function removes the finished to-do from the AsyncStorage, and replaces it with "done" at the start of the id.
    handleFinishedTodo(list){
        const id = list[0]; const text = list[1]; const date = list[2]; const done = list[3];
        const AsyncList = [text,date,done];

        AsyncStorage.removeItem(id.toString());

        this.storeTodo("done"+id.toString(), JSON.stringify(AsyncList));
        this.setState({finishedTodoList:list});
    }

    //This function is triggered when the user press the button "Show finished todos"
    //It updates the list being to shown to the user with the preferred list (todos or finished todos)
    handleShowFinishedTodos(){
        this.setState({showFinishedTodos:!this.state.showFinishedTodos});
        if(this.state.showFinishedTodos){
            this.updateShowList(this.state.finishedTodoList)
        }
        else{
            this.updateShowList(this.state.todoList)
        }
    }

    //updateSortedList immediately updates the list when the date is changed.
    //if a to-do with an older date than another to-do is changed, the to-do will move above
    updateSortedList(list){
        const updateList = this.state.todoList.slice();
        updateList.map((item) => {
            //checks if the to-do which is changed matches an todonr in the todolist
            if(item.todoNr === list[0]){
                //update the todoDate to the new one
                item.todoDate = list[2];
            }
        });

        //The list is being sat to both to-do- and showlist for immediate change on the screen.
        this.updateTodoList(updateList);
        this.updateShowList(updateList);
    }

    //deleteTodo deletes an to-do from the todolist.
    //The function gets an id as parameter and, creates a new temporarily list, and adds each element from the todolist except
    //the element with the given id
    deleteTodo = (id) => {
        const deleteList = [];
        for(let i = 0; i < this.state.todoList.length; i++){
            if(this.state.todoList[i].todoNr !== id){
                deleteList.push(this.state.todoList[i]);
            }
        }
        //The list is being sat to both to-do- and showlist for immediate change on the screen.
        this.updateTodoList(deleteList);
        this.updateShowList(deleteList);

        //The item is also removed from the AsyncStorage
        AsyncStorage.removeItem(id.toString());
    };

    //This function sorts the current list by date
    sortByDate(todoList){
        const todosWithDate = [];
        const todosWithoutDate = [];

        todoList.map((item) => {
            if(item.todoDate !== ""){
                todosWithDate.push(item);
            }
            else{
                todosWithoutDate.push(item);
            }
        });

        todosWithDate.sort(function(a,b){
            if(a.todoDate !== "" && b.todoDate !== ""){
                let c = new Date(a.todoDate);
                let d = new Date(b.todoDate);
                return c-d;
            }
        });
        return todosWithDate.concat(todosWithoutDate);
    };

    //This is the list being shown to the user, and this function updates the list.
    updateShowList(list){
        const sortedByDate = this.sortByDate(list);
        this.setState({showList:sortedByDate});
    }

    //Updates the todolist, so the user get immediate update.
    updateTodoList(list){
        const sortedByDate = this.sortByDate(list);
        this.setState({todoList:sortedByDate});
    }

    //Function to store values in the AsyncStorage
    storeTodo = async (id, data) => {
        try {
            await AsyncStorage.setItem(id, data);
        } catch (error) {
            throw error;
        }
    };

    render() {
        return (
            <View style={styles.container}>
                <View style={styles.toolbar}>
                    {/*Textinput that automatically gets focused when the user enters the todopage.*/}
                    {/*When text is written, the todoTextstate gets changed*/}
                    <TextInput
                        placeholder="write something here"
                        ref={this.myTextInput}
                        style={styles.textInput}
                        onChangeText={(text) => this.setState({todoText : text})}
                        value={this.state.todoText}
                    />
                    {/*Creates a to-do with the given text written in the textinput, and the current todonr*/}
                    <TouchableOpacity style={styles.toolbarAddBtn} onPress={() => this.addTodo(this.state.currTodoNr)}>
                        <Text style={styles.toolbarAddBtnText}>Add</Text>
                    </TouchableOpacity>
                </View>
                <View  style={styles.finishedTodosBtn}>
                    <Button
                        onPress={() => this.handleShowFinishedTodos()}
                        title={this.state.showFinishedTodos ? "Show finished todos" : "Show todos"}
                    />
                </View>
                <View style={{flex:1}}>
                    {/*Scrollview so the user could scroll if the amount of todos become large*/}
                    {/*Maps through the todolist (which is a prop from app.js), and creates a listitem for each to-do*/}
                    {<ScrollView>
                        {this.state.showList.map((element) =>
                        <ListItem handleFinishedTodo = {this.handleFinishedTodo} updateSortedList = {this.updateSortedList} deleteTodo = {this.deleteTodo} name = {element.todoText} key = {element.key} done = {element.done} date = {element.todoDate} todoNr = {element.todoNr}/>
                    )}
                    </ScrollView>}
                </View>
            </View>
        );
    }
};