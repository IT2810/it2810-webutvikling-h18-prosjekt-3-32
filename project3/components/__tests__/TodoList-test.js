import React from "react";
import TodoList from "../TodoList";

import ShallowRenderer from "react-test-renderer/shallow";
import renderer from "react-test-renderer";
import MockAsyncStorage from "mock-async-storage";

import { AsyncStorage as storage } from "react-native";

//Snapshot testing of TodoList.js
test("TodoList renders correctly!", () => {
   const tree = renderer.create(<TodoList/>).toJSON();
   expect(tree).toMatchSnapshot();
});

//Test to check if the function addTodo works
test("testing addTodo in TodoList", () => {
    //Creates instance of TodoList
    const testTodoListComponent = <TodoList/>;
    const todoComponent = renderer.create(testTodoListComponent).getInstance();

    //Setting state and triggers addTodo with todonr: 1
    todoComponent.setState({todoText: "test"});
    todoComponent.addTodo(1);

    //Testlist to test state.todoList upon
    const testList = [{
        key: 1,
        todoNr: "todo1",
        todoText: "test",
        todoDate: "",
        done: false,
    }];

    //If addTodo did add the to-do, the todolist will be equal to testList
    expect(todoComponent.state.todoList).toEqual(testList);
});

//Test to check if the function deleteTodo works
test("testing deleteTodo in Todolist", () => {
    //Creates instance of TodoList
    const testTodoListComponent = <TodoList/>;
    const todoComponent = renderer.create(testTodoListComponent).getInstance();

    //Setting state.todoList to be only this testList
    const testList = [{
        key: 1,
        todoNr: "todo1",
        todoText: "test",
        todoDate: "",
        done: false,
    }];
    todoComponent.setState({todoList:testList});

    //Triggers deleTodo
    todoComponent.deleteTodo("todo1");

    const emptyList = [];

    //Checks is the testList is gone and todoList is empty
    expect(todoComponent.state.todoList).toEqual(emptyList);
});

//Test to check if the function updateSortedList works
test("check update Todo with date", () =>{
    //Creates instance of TodoList
    const testTodoListComponent = <TodoList/>;
    const todoComponent = renderer.create(testTodoListComponent).getInstance();

    todoComponent.setState({todoText: "test"});
    todoComponent.addTodo(1);

    //Checking with a list with date
    const testListWithDate = ["todo1", "test", "2018-12-24"];

    //Updating the todolist with the function updateSortedList
    todoComponent.updateSortedList(testListWithDate);

    const testList = [{
        key: 1,
        todoNr: "todo1",
        todoText: "test",
        todoDate: "2018-12-24",
        done: false,
    }];

    //Checks whether the date was added
    expect(todoComponent.state.todoList).toEqual(testList);
});

//Test to check if the function sortByDate works
test("check if the function sortByDate works", () => {
    //Creates instance of TodoList
    const testTodoListComponent = <TodoList/>;
    const todoComponent = renderer.create(testTodoListComponent).getInstance();

    //Creates an unsorted list
    const unsortedListByDate = [
        {key : 1, todoNr : "todo1", todoText: "test1", todoDate: "9999-12-24"},
        {key : 2, todoNr : "todo2", todoText: "test2", todoDate: "2000-12-24"}
    ];

    //Runs sortByDate which also triggers sortByDate
    const sortedList = todoComponent.sortByDate(unsortedListByDate);


    //Creates a sorted list
    const sortedListByDate = [
        {key : 2, todoNr : "todo2", todoText: "test2", todoDate: "2000-12-24"},
        {key : 1, todoNr : "todo1", todoText: "test1", todoDate: "9999-12-24"}
    ];

    //Checks if the state is the sorted list
    expect(sortedList).toEqual(sortedListByDate);
});

//Test to check if the function handleFinishedTodo works
test("check if handleFinishedTodo works", async () => {
    //Creates instance of TodoList
    const testTodoListComponent = <TodoList/>;
    const todoComponent = renderer.create(testTodoListComponent).getInstance();

    //Creates a finished todolist
    let finishedList = [{
        key:1,
        todoNr:"todo1",
        todoText:"test",
        todoDate:"2018-12-24",
        done:true
    }];

    //Triggers handleFinishedTodo with the finishedList as parameter
    await todoComponent.handleFinishedTodo(finishedList);

    //Checks if the state is updated
    expect(todoComponent.state.finishedTodoList).toEqual(finishedList);

});

test("check if handleShowFinishedTodos works", () => {
    //Creates instance of TodoList
    const testTodoListComponent = <TodoList/>;
    const todoComponent = renderer.create(testTodoListComponent).getInstance();

    //Creates a finished todolist
    let finishedList = [{
        key:1,
        todoNr:"todo1",
        todoText:"test",
        todoDate:"2018-12-24",
        done:true
    }];

    //Setting the state to be true for the first expect
    todoComponent.setState({showFinishedTodos:false, finishedTodoList:finishedList});

    //Triggers handleShowFinishedTodos which will enter the if-statement
    todoComponent.handleShowFinishedTodos();

    //Checks if the state for finishedTodoList is updated
    expect(todoComponent.state.showList).toEqual(finishedList);

    //Creates an unfinished todolist
    let unfinishedList = [{
        key:1,
        todoNr:"todo1",
        todoText:"test",
        todoDate:"2018-12-24",
        done:false
    }];

    //Setting the state to be false for the second expect
    todoComponent.setState({showFinishedTodos:true, todoList:unfinishedList});

    //Triggers handleShowFinishedTodos which will enter the else-statement
    todoComponent.handleShowFinishedTodos();

    //Checks if the state todoList is updated
    expect(todoComponent.state.showList).toEqual(unfinishedList);
});

//---Testing functions in Render() (onPress, onChangeText..)---//

//Test to check if onChangeValue at TextInput works
test("test onChangeText at textInput", () => {
    const mockOnChangeValue = jest.fn();

    //Creates a shallowRenderer to get the renderoutput of TodoList
    const shallow = new ShallowRenderer();
    shallow.render(<TodoList/>);
    const result = shallow.getRenderOutput();

    //Navigating to find the onChangeText at textinput and triggers it
    result.props.children[0].props.children[0].props.onChangeText(mockOnChangeValue("test"));

    //Checks if the onChangeText was triggered one time
    expect(mockOnChangeValue).toHaveBeenCalledTimes(1);
});

//Test to check if onPress at Button works
test("test onPress at Button", () => {
    const mockOnPress = jest.fn();

    //Creates a shallowRenderer to get the renderoutput of TodoList
    const shallow = new ShallowRenderer();
    shallow.render(<TodoList/>);
    const result = shallow.getRenderOutput();

    //Navigating to find the onPress at Button and triggers it
    result.props.children[1].props.children.props.onPress(mockOnPress(true));

    //Checks if the onChangeText was triggered one time
    expect(mockOnPress).toHaveBeenCalledTimes(1);
});

//---Testing AsyncStorage---//

test("Functions using AsyncStorage", async () => {
    //Setting up mock Async
    const asyncMock = () => {
        const mockImpl = new MockAsyncStorage();
        jest.mock("AsyncStorage", () => mockImpl);
    };

    //Creates instance of TodoList
    const testTodoListComponent = <TodoList/>;
    const todoComponent = renderer.create(testTodoListComponent).getInstance();

    //Creates a testList
    const testList = ["todo1", "test", "", false];

    //Initialize the mock for AsyncStorage
    asyncMock();

    //--Testing StoreTodo in TodoList (Storing in AsyncStorage)--//
    const testAsyncList = [testList[1], testList[2], testList[3]];

    //Triggers storeTodo in TodoList
    await todoComponent.storeTodo(testList[0].toString(), JSON.stringify(testAsyncList));
    const value = await storage.getItem(testList[0].toString());
    const parsedValue = JSON.parse(value);

    expect(parsedValue).toEqual(testAsyncList);

    //--Testing removeTodo in TodoList (Removing item in AsyncStorage)--//

    //Triggers removeTodo in TodoList, which will remove the item in the mockAsyncStorage
    await todoComponent.removeTodo(testList[0].toString());
    const value1 = await storage.getItem(testList[0].toString());

    //Will be undefined since there isn't anything there.
    expect(value1).toBeUndefined();
});