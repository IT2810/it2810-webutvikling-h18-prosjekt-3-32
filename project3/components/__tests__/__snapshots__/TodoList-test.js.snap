// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders correctly! 1`] = `
<View
  style={
    Object {
      "borderRadius": 4,
      "height": "100%",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f7f7f7",
        "flexDirection": "row",
        "height": "13%",
        "justifyContent": "space-around",
        "width": "100%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="write something here"
      style={
        Object {
          "color": "#a0a0a0",
          "fontSize": 18,
          "height": "80%",
          "width": "70%",
        }
      }
      value=""
    />
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#5cacec",
          "borderRadius": 25,
          "borderWidth": 1,
          "height": 50,
          "justifyContent": "center",
          "marginRight": 10,
          "opacity": 1,
          "width": 50,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#5cacec",
            "fontSize": 15,
          }
        }
      >
        Add
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomWidth": 0.5,
        "borderTopWidth": 0.5,
        "height": "7%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Show finished todos
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;
